pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,246 @@
+	"github.com/opencontainers/selinux/go-selinux"
+	"github.com/rancher/k3s-upgrade/pkg/types"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func GetProcessPID() (int, error) {
+	processes, err := ps.Processes()
+	if err != nil {
+		return 0, err
+	}
+	for _, proc := range processes {
+		if strings.HasSuffix(proc.Executable(), "k3s") {
+			if strings.Contains(proc.Executable(), "init") || strings.Contains(proc.Executable(), "channelserver") {
+				continue
+			}
+			return proc.Pid(), nil
+		}
+	}
+	return -1, fmt.Errorf("k3s process is not running")
+}
+
+func GetBinPath(pid int) (string, error) {
+	_, err := os.FindProcess(pid)
+	if err != nil {
+		return "", fmt.Errorf(" process is not found")
+	}
+	if pid == 1 {
+		return "/bin/k3s", nil
+	}
+	binPath, err := ioutil.ReadFile(path.Join("/host/proc", strconv.Itoa(pid), "cmdline"))
+	if fileExists(string(binPath)) {
+		return string(binPath), nil
+	}
+	return "", fmt.Errorf(" failed to fetch the k3s binary path from process %d", pid)
+}
+
+func TermProcess(pid int) error {
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		fmt.Errorf(" process is not found")
+	}
+	return proc.Signal(syscall.SIGTERM)
+}
+
+func MatchSelector(hostPath, configPath string, preservedEntries []string) error {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	conf, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return err
+	}
+	nodeClient := kubernetes.NewForConfigOrDie(conf).CoreV1().Nodes()
+	hostname := os.Getenv("SYSTEM_UPGRADE_NODE_NAME")
+	if hostname == "" {
+		hostname, err = os.Hostname()
+		if err != nil {
+			return err
+		}
+	}
+	host, err := nodeClient.Get(context.Background(), hostname, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	hostLabels := host.GetLabels()
+	secretfiles, err := ioutil.ReadDir(configPath)
+	if err != nil {
+		return err
+	}
+	var secret types.Secret
+	secretfiles = sortFileByName(secretfiles)
+	for _, secretFile := range secretfiles {
+		fileContent, err := ioutil.ReadFile(secretFile.Name())
+		if err != nil {
+			return err
+		}
+		yaml.Unmarshal(fileContent, secret)
+		selector, err := v1.LabelSelectorAsSelector(&secret.Selector)
+		if err != nil {
+			return err
+		}
+		if selector.Matches(labels.Set(hostLabels)) {
+			//check if a config file is present on the node
+			return PlaceConfigFile(hostPath, secret.Config, preservedEntries)
+
+		}
+	}
+	return nil
+}
+
+func PlaceConfigFile(hostPath string, config map[string]interface{}, preservedEntries []string) error {
+	preserved := make(map[string]interface{})
+	if fileExists(hostPath) {
```

Comment:
Please see my comment above on the `fileExists()` function. I don't think we want to potentially swallow `os.Stat()` errors here, we should return such to callers of `PlaceConfigFile()`